File: Stacks/394. Decode String/394. Decode String.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function (s) {
  let stack = [];

  for (const char of s) {
    if (char !== "]") stack.push(char);
    else {
      let currString = "";
      while (stack.at(-1) !== "[") {
        currString = stack.pop() + currString;
      }
      stack.pop();

      let num = "";
      while (stack && !isNaN(stack.at(-1))) {
        num = stack.pop() + num;
      }

      stack.push(currString.repeat(Number(num)));
    }
  }

  return stack.join("");
};

// Time: O(L)
// Space: O(L)
// Where L is the length of the decoded string

File: Stacks/394. Decode String/394. Decode String.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []

        for char in s:
            if char != "]":
                stack.append(char)
            else:
                curr_string = ""
                while stack[-1] != "[":
                    curr_string = stack.pop() + curr_string
                stack.pop()
                
                num = ""
                while stack and stack[-1].isdigit():
                    num = stack.pop() + num

                stack.append(int(num) * curr_string)
            
        return "".join(stack)
            
# Time: O(L)
# Space: O(L)
# Where L is the length of the decided string